File: Avail/src/com/avail/builder/ModuleRoots.java
/**
 * ModuleRoots.java
 * Copyright © 1993-2013, Mark van Gulik and Todd L Smith.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * * Redistributions of source code must retain the above copyright notice, this
 *   list of conditions and the following disclaimer.
 *
 * * Redistributions in binary form must reproduce the above copyright notice,
 *   this list of conditions and the following disclaimer in the documentation
 *   and/or other materials provided with the distribution.
 *
 * * Neither the name of the copyright holder nor the names of the contributors
 *   may be used to endorse or promote products derived from this software
 *   without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

package com.avail.builder;

import static com.avail.persistence.IndexedRepositoryManager.isIndexedRepositoryFile;
import java.io.*;
import java.util.*;
import com.avail.annotations.*;
import com.avail.descriptor.ModuleDescriptor;

/**
 * {@code ModuleRoots} encapsulates the Avail {@linkplain ModuleDescriptor
 * module} path in both composed and decomposed forms. The Avail module path
 * specifies bindings between <em>logical root names</em> and {@linkplain
 * ModuleRoot locations} of Avail modules. A logical root name should typically
 * belong to a vendor of Avail modules, ergo a domain name or registered
 * trademark suffices nicely.
 *
 * <p>The format of an Avail module path is described by the following
 * simple grammar:</p>
 *
 * <pre>
 * modulePath ::= binding ++ ";" ;
 * binding ::= root "=" directory ;
 * root ::= [^=;]+ ;
 * directory ::= [^;]+ ;
 * </pre>
 *
 * @author Todd L Smith &lt;todd@availlang.org&gt;
 */
@ThreadSafe
public final class ModuleRoots
{
	/** The Avail {@linkplain ModuleDescriptor module} path. */
	private final String modulePath;

	/**
	 * Answer the Avail {@linkplain ModuleDescriptor module} path.
	 *
	 * @return The Avail {@linkplain ModuleDescriptor module} path.
	 */
	public String modulePath ()
	{
		return modulePath;
	}

	/**
	 * A {@linkplain Map map} from logical root names to {@linkplain ModuleRoot
	 * module root}s.
	 */
	private final Map<String, ModuleRoot> rootMap =
		new LinkedHashMap<>();

	/**
	 * Parse the Avail {@linkplain ModuleDescriptor module} path into a
	 * {@linkplain Map map} of logical root names to {@linkplain ModuleRoot
	 * module root}s.
	 *
	 * @throws IllegalArgumentException
	 *         If any component of the Avail {@linkplain ModuleDescriptor
	 *         module} path is invalid.
	 */
	private void parseAvailModulePath ()
		throws IllegalArgumentException
	{
		// Root definitions are separated by semicolons.
		for (final String component : modulePath.split(";"))
		{
			// An equals separates the root name from its paths.
			final String[] binding = component.split("=");
			if (binding.length != 2)
			{
				throw new IllegalArgumentException();
			}

			// A comma separates the repository path from the source directory
			// path.
			final String rootName = binding[0];
			final String[] paths = binding[1].split(",");
			if (paths.length > 2)
			{
				throw new IllegalArgumentException();
			}

			// All paths must be absolute.
			for (final String path : paths)
			{
				final File file = new File(path);
				if (!file.isAbsolute())
				{
					throw new IllegalArgumentException();
				}
			}

			// If only one path is supplied, then it must reference a valid
			// repository.
			final File repositoryFile = new File(paths[0]);
			try
			{
				if (paths.length == 1
					&& !isIndexedRepositoryFile(repositoryFile))
				{
					throw new IllegalArgumentException();
				}
			}
			catch (final IOException e)
			{
				throw new IllegalArgumentException(e);
			}

			// If two paths are provided, then the first path need not reference
			// an existing file. The second path, however, must reference a
			// directory.
			final File sourceDirectory =
				paths.length == 2
				? new File(paths[1])
				: null;
			if (sourceDirectory != null && !sourceDirectory.isDirectory())
			{
				throw new IllegalArgumentException();
			}

			rootMap.put(
				rootName,
				new ModuleRoot(rootName, repositoryFile, sourceDirectory));
		}
	}

	/**
	 * Answer the logical root names in the order that they are specified in
	 * the Avail {@linkplain ModuleDescriptor module} path.
	 *
	 * @return The logical root names.
	 */
	public Set<String> rootNames ()
	{
		return Collections.unmodifiableSet(rootMap.keySet());
	}

	/**
	 * Answer the {@linkplain ModuleRoot module roots} in the order that they
	 * are specified in the Avail {@linkplain ModuleDescriptor module} path.
	 *
	 * @return The module roots.
	 */
	public Set<ModuleRoot> roots ()
	{
		final Set<ModuleRoot> roots = new LinkedHashSet<>();
		for (final Map.Entry<String, ModuleRoot> entry : rootMap.entrySet())
		{
			roots.add(entry.getValue());
		}
		return roots;
	}

	/**
	 * Answer the {@linkplain ModuleRoot module root} bound to the specified
	 * logical root name.
	 *
	 * @param rootName
	 *        A logical root name, typically something owned by a vendor of
	 *        Avail {@linkplain ModuleDescriptor modules}.
	 * @return The module root, or {@code null} if no such binding exists.
	 */
	public @Nullable ModuleRoot moduleRootFor (final String rootName)
	{
		return rootMap.get(rootName);
	}

	/**
	 * Construct a new {@link ModuleRoots} from the specified Avail {@linkplain
	 * ModuleDescriptor module} path.
	 *
	 * @param modulePath
	 *        An Avail {@linkplain ModuleDescriptor module} path.
	 * @throws IllegalArgumentException
	 *         If the Avail {@linkplain ModuleDescriptor module} path is
	 *         malformed.
	 */
	public ModuleRoots (final String modulePath)
	{
		this.modulePath = modulePath;
		parseAvailModulePath();
	}
}

File: src/main/kotlin/com/avail/builder/ModuleRoots.kt
/*
 * ModuleRoots.kt
 * Copyright © 1993-2020, The Avail Foundation, LLC.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * * Redistributions of source code must retain the above copyright notice, this
 *   list of conditions and the following disclaimer.
 *
 * * Redistributions in binary form must reproduce the above copyright notice,
 *   this list of conditions and the following disclaimer in the documentation
 *   and/or other materials provided with the distribution.
 *
 * * Neither the name of the copyright holder nor the names of the contributors
 *   may be used to endorse or promote products derived from this software
 *   without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

package com.avail.builder

import com.avail.annotations.ThreadSafe
import com.avail.descriptor.module.ModuleDescriptor
import com.avail.files.FileManager
import com.avail.persistence.cache.Repositories
import com.avail.resolver.ModuleRootResolver
import com.avail.resolver.ModuleRootResolverRegistry
import com.avail.utility.json.JSONWriter
import java.net.URI
import java.util.Collections.unmodifiableSet
import java.util.concurrent.atomic.AtomicInteger
import java.util.concurrent.locks.ReentrantLock

/**
 * `ModuleRoots` encapsulates the Avail [module][ModuleDescriptor] path. The
 * Avail module path specifies bindings between *logical root names* and
 * [locations][ModuleRoot] of Avail modules. A logical root name should
 * typically belong to a vendor of Avail modules, ergo a domain name or
 * registered trademark suffices nicely.
 *
 * The format of an Avail module path is described by the following simple
 * grammar:
 *
 * ```
 * modulePath ::= binding ++ ";" ;
 * binding ::= logicalRoot "=" objectRepository ("," sourceDirectory) ;
 * logicalRoot ::= [^=;]+ ;
 * objectRepository ::= [^;]+ ;
 * sourceDirectory ::= [^;]+ ;
 * ```
 *
 * `logicalRoot` represents a logical root name. `objectRepository` represents
 * the absolute path of a binary module repository. `sourceDirectory` represents
 * the absolute path of a package, i.e., a directory containing source modules,
 * and may be sometimes be omitted (e.g., when compilation is not required).
 *
 * @author Todd L Smith &lt;todd@availlang.org&gt;
 * @author Richard Arriaga &lt;rich@availlang.org&gt;
 *
 * @property fileManager
 *   The associated [FileManager].
 *
 * @constructor
 *
 * Construct a new `ModuleRoots` from the specified Avail roots path.
 *
 * @param fileManager
 *   The associated [FileManager].
 * @param modulePath
 *   An Avail [module][ModuleDescriptor] path.
 * @param withFailures
 *   A lambda that accepts [List] of the string [ModuleRoot] [URI]s that failed
 *   to [resolve][ModuleRootResolver.resolve].
 * @throws IllegalArgumentException
 *   If the Avail [module][ModuleDescriptor] path is malformed.
 */
@ThreadSafe
class ModuleRoots constructor(
	val fileManager: FileManager,
	modulePath: String,
	withFailures: (List<String>) -> Unit) : Iterable<ModuleRoot>
{
	/**
	 * A [map][Map] from logical root names to [module&#32;root][ModuleRoot]s.
	 */
	private val rootMap = LinkedHashMap<String, ModuleRoot>()

	/**
	 * The Avail [module][ModuleDescriptor] path.
	 */
	val modulePath: String by lazy {
		val builder = StringBuilder(200)
		var first = true
		for ((_, root) in rootMap)
		{
			if (!first)
			{
				builder.append(";")
			}
			builder.append(root.name)
			builder.append("=")
			builder.append(root.repository.fileName.path)
			val resolver = root.resolver
			builder.append(",")
			builder.append(resolver.uri.toString())
			first = false
		}
		builder.toString()
	}

	/**
	 * Parse the Avail [module][ModuleDescriptor] path into a [map][Map] of
	 * logical root names to [module&#32;root][ModuleRoot]s.
	 *
	 * @param modulePath
	 *   The module roots path string.
	 * @param withFailures
	 *   A lambda that accepts [List] of the string [ModuleRoot] [URI]s that failed
	 *   to [resolve][ModuleRootResolver.resolve].
	 * @throws IllegalArgumentException
	 *   If any component of the Avail [module][ModuleDescriptor] path is
	 *   invalid.
	 */
	private fun parseAvailModulePathThen(
		modulePath: String, withFailures: (List<String>) -> Unit)
	{
		clearRoots()
		// Root definitions are separated by semicolons.
		val components =
			if (modulePath.isEmpty())
			{
				listOf()
			}
			else
			{
				modulePath.split(";")
			}
		val workCount = AtomicInteger(components.size)
		val failures = mutableListOf<String>()
		val lock = ReentrantLock()
		for (component in components)
		{
			// An equals separates the root name from its paths.
			val binding = component.split("=")
			require(binding.size == 2) {
				"Bad module root location setting: $component"
			}

			// A comma separates the repository path from the source directory
			// path.
			val rootName = binding[0]
			val location = binding[1]

			val rootUri = URI(location)
			val resolver =
				ModuleRootResolverRegistry.createResolver(
					rootName,
					rootUri,
					fileManager)

			resolver.resolve(
				{
					synchronized(lock)
					{
						addRoot(resolver.moduleRoot)
						if (workCount.decrementAndGet() == 0)
						{
							withFailures(failures)
						}
					}
				}
			) { code, ex ->
				val message =
					"$code: Could not resolve module root, $rootName ($rootUri)"
				synchronized(lock)
				{
					failures.add(message)
				}
				System.err.println(message)
				ex?.printStackTrace()
				if (workCount.decrementAndGet() == 0)
				{
					withFailures(failures)
				}
			}
		}
	}

	/**
	 * Clear the [root&#32;map][rootMap].
	 */
	fun clearRoots() = rootMap.clear()

	/**
	 * Add a [root][ModuleRoot] to the [root&#32;map][rootMap].
	 *
	 * @param root
	 *   The root.
	 */
	fun addRoot(root: ModuleRoot)
	{
		rootMap[root.name] = root
		Repositories.addRepository(root)
	}

	/**
	 * Fully remove the provided [ModuleRoot.name].
	 *
	 * @param name
	 *   The name of the root to remove.
	 */
	fun removeRoot (name: String)
	{
		rootMap.remove(name)?.let {
			Repositories.deleteRepository(name)
		}
	}

	/**
	 * The logical root names in the order that they are specified in the Avail
	 * [module][ModuleDescriptor] path.
	 */
	val rootNames: Set<String> get() = unmodifiableSet(rootMap.keys)

	/**
	 * The [module&#32;roots][ModuleRoot] in the order that they are specified
	 * in the Avail [module][ModuleDescriptor] path.
	 */
	val roots get () = rootMap.values.toSet()

	override fun iterator () = roots.toSet().iterator()

	/**
	 * Answer the [module&#32;root][ModuleRoot] bound to the specified logical
	 * root name.
	 *
	 * @param rootName
	 *   A logical root name, typically something owned by a vendor of Avail
	 *   [modules][ModuleDescriptor].
	 * @return
	 *   The module root, or `null` if no such binding exists.
	 */
	fun moduleRootFor(rootName: String): ModuleRoot? = rootMap[rootName]

	init
	{
		parseAvailModulePathThen(modulePath, withFailures)
	}

	/**
	 * Write a JSON encoding of the module roots to the specified [JSONWriter].
	 *
	 * @param writer
	 *   A `JSONWriter`.
	 */
	fun writeOn(writer: JSONWriter)
	{
		writer.writeArray {
			roots.forEach { root -> write(root.name) }
		}
	}

	/**
	 * Write a JSON object whose fields are the module roots and whose values
	 * are [JSON&#32;arrays][ModuleRoot.writePathsOn] containing path
	 * information.
	 *
	 * @param writer
	 *   A [JSONWriter].
	 */
	fun writePathsOn(writer: JSONWriter)
	{
		writer.writeArray {
			roots.forEach { root ->
				at(root.name) { root.writePathsOn(writer) }
			}
		}
	}
}
